src/services/videoService.js

```diff
@@ -1,12 +1,97 @@
-      description: video.snippet.description,
-    };
+function initializeYoutubeApi(fn) {
+  // initialize gapi
+  window.gapi.load('client:auth2', async () => {
+    await window.gapi.auth2.init({
+      client_id:
+        '537606572960-o8ms8f2g27kepgh2fmd14m0bgsmsh6og.apps.googleusercontent.com',
+    });
+    // to initialize youtube we need to set the api key first
+    window.gapi.client.setApiKey('AIzaSyCykzCobNVjS-XeWrXWwPH06nLtq7Assis');
+    await window.gapi.client.load(
+      'https://www.googleapis.com/discovery/v1/apis/youtube/v3/rest'
+    );
+
+    await fn();
+  });
+}
+
+export function getVideos(setVideos, keyword) {
+  initializeYoutubeApi(async () => {
+    // after initializing we can make the api call
+    // call youtube search
+    const response = await window.gapi.client.youtube.search.list({
+      part: ['snippet'],
+      maxResults: 25,
+      q: keyword,
+    });
+
+    const videos = response.result.items
+      .map((video) => {
+        return {
+          etag: video.etag,
+          videoId: video.id.videoId,
+          title: video.snippet.title,
+          imageUrl: video.snippet.thumbnails.medium.url,
+          description: video.snippet.description,
+        };
+      })
+      .filter((video) => video.videoId);
+
+    // use the setVideos function passed by the component to set the videos to the state
+    setVideos(videos);
+  });
+}
+
+export function getRelatedVideos(setRelatedVideos, videoId) {
+  initializeYoutubeApi(async () => {
+    // call youtube search
+    const response = await window.gapi.client.youtube.search.list({
+      part: ['snippet'],
+      maxResults: 25,
+      relatedToVideoId: videoId,
+    });
+
+    const videos = response.result.items
+      .map((video) => {
+        return {
+          etag: video.etag,
+          videoId: video.id.videoId,
+          title: video.snippet.title,
+          imageUrl: video.snippet.thumbnails.medium.url,
+          description: video.snippet.description,
+        };
+      })
+      .filter((video) => video.videoId);
+
+    // use the setVideos function passed by the component to set the videos to the state
+    setRelatedVideos(videos);
+  });
+}
+
+export function loadVideo(setVideo, videoId) {
+  console.log(videoId);
+  initializeYoutubeApi(async () => {
+    // call youtube videos list
+    const response = await window.gapi.client.youtube.videos.list({
+      part: ['snippet,contentDetails,statistics'],
+      id: [videoId],
+    });
+
+    const videos = response.result.items
+      .map((video) => {
+        return {
+          etag: video.etag,
+          videoId,
+          title: video.snippet.title,
+          imageUrl: video.snippet.thumbnails.medium.url,
+          description: video.snippet.description,
+        };
+      })
+      .filter((video) => video.videoId);
+
+    const video = videos[0];
+
+    // use the setVideo function passed by the component to set the video to the state
+    setVideo(video);
```

Comment:
You're kind of doing what a custom hook would do. It receives some data, handles it, and returns a result. Try abstracting this to a custom hook you could use in your components.

Reply:
For testing the custom hook take a look at:

Basic hook testing:
https://react-hooks-testing-library.com/usage/basic-hooks

For async calls or updates to the hook:
https://react-hooks-testing-library.com/usage/advanced-hooks#async

You're required to install an additional package for hook testing, check the installation guide for it. Let me know if you have any questions.

Reply:
I'm going to try to use fetch instead of gapi and maybe the custom hook won't be necessary